### PR TITLE
Fix case where # occurs after > in JSX expression

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Bug #4207

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -929,6 +935,13 @@ namespace ts {
             return false;
         }
 
+        function parseOptionalWithoutAdvancing(t: SyntaxKind): boolean {
+            if (token === t) {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: `return token === t`?

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -3308,14 +3321,14 @@ namespace ts {
             return finishNode(node);
         }
 
-        function parseJsxElementOrSelfClosingElement(): JsxElement|JsxSelfClosingElement {
-            let opening = parseJsxOpeningOrSelfClosingElement();
+        function parseJsxElementOrSelfClosingElement(inExpressionContext: boolean): JsxElement|JsxSelfClosingElement {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Space around `|`

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -3336,9 +3349,9 @@ namespace ts {
                 case SyntaxKind.JsxText:
                     return parseJsxText();
                 case SyntaxKind.OpenBraceToken:
-                    return parseJsxExpression();
+                    return parseJsxExpression(/*inExpression*/false);
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Space before `false`.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -3336,9 +3349,9 @@ namespace ts {
                 case SyntaxKind.JsxText:
                     return parseJsxText();
                 case SyntaxKind.OpenBraceToken:
-                    return parseJsxExpression();
+                    return parseJsxExpression(/*inExpression*/false);
                 case SyntaxKind.LessThanToken:
-                    return parseJsxElementOrSelfClosingElement();
+                    return parseJsxElementOrSelfClosingElement(/*inExpression*/false);
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Space before `false`.

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -913,9 +917,11 @@ namespace ts {
             return token > SyntaxKind.LastReservedWord;
         }
 
-        function parseExpected(kind: SyntaxKind, diagnosticMessage?: DiagnosticMessage): boolean {
+        function parseExpected(kind: SyntaxKind, diagnosticMessage?: DiagnosticMessage, advance = true): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: advance -> shouldAdvance

File: src/compiler/parser.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: inExpressionContext

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -3406,14 +3429,20 @@ namespace ts {
...
-        function parseJsxExpression(): JsxExpression {
+        function parseJsxExpression(inExpressionContext: boolean): JsxExpression {
             let node = <JsxExpression>createNode(SyntaxKind.JsxExpression);
 
             parseExpected(SyntaxKind.OpenBraceToken);
             if (token !== SyntaxKind.CloseBraceToken) {
                 node.expression = parseExpression();
             }
-            parseExpected(SyntaxKind.CloseBraceToken);
+            if(inExpressionContext) {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Space after `if`

File: src/compiler/parser.ts [hunk 7]
```diff
@@ -3336,9 +3342,9 @@ namespace ts {
                 case SyntaxKind.JsxText:
                     return parseJsxText();
                 case SyntaxKind.OpenBraceToken:
-                    return parseJsxExpression();
+                    return parseJsxExpression(/*inExpression*/ false);
                 case SyntaxKind.LessThanToken:
-                    return parseJsxElementOrSelfClosingElement();
+                    return parseJsxElementOrSelfClosingElement(/*inExpression*/ false);
```
Comment (yuit) [inline src/compiler/parser.ts]: ```/inExpressionContext*/```

File: src/compiler/parser.ts [hunk 8]
```diff
@@ -3336,9 +3342,9 @@ namespace ts {
                 case SyntaxKind.JsxText:
                     return parseJsxText();
                 case SyntaxKind.OpenBraceToken:
-                    return parseJsxExpression();
+                    return parseJsxExpression(/*inExpression*/ false);
```
Comment (yuit) [inline src/compiler/parser.ts]: `/*inExpressionContext*/`

**Pull request merged**